<launch>

        <!-- Controller Node -->
        <node pkg="quad_control" name="plot_nodeasas" type="plot_experiment.py"/>

        <!-- Planner Node -->
        <node pkg="quad_control" name="planner_node" type="planner_node.py"/>
        <!-- Collision avoidance Node -->
        <node pkg="quad_control" name="collision_avoidance_node" type="collision_avoidance_node.py"/>
        <!-- Magnitude control Node -->
        <node pkg="quad_control" name="magnitude_control_node" type="magnitude_control_node.py"/>

        <!-- Default Mission -->
        <param name="MissionDefault" type="str" value="IrisRealTrajectoryTracking"/>

        <!-- Default Yaw Controller -->
        <param name="YawControllerDefault" type="str" value="SimpleTrackingYawController"/>

        <!-- Default Trajectory -->
        <param name="TrajectoryDefault" type="str" value="StayAtRest"/>
        <param name="trajectry_offset_x" type="double" value="0.0"/>
        <param name="trajectry_offset_y" type="double" value="0.0"/>
        <param name="trajectry_offset_z" type="double" value="0.0"/>


        <!-- Default Yaw Controller -->
        <param name="ControllerDefault" type="str" value="SimplePIDController"/>
        <param name="DIControllerDefault" type="str" value="ComponentWise3DDIC"/>


        <!-- Controller Node -->
        <node pkg="quad_control" name="controller_Iris1" type="cycle_quad_control_mission.py"  output="screen"/>

        <!-- Controller parameters -->
        <param name="gravity_ctr" value="9.81"/>
        <param name="mass_quad_ctr" value="1.442"/>
        <param name="Throttle_neutral_ctr" value="1484.0"/>  
        <param name="ACRO_RP_P_ctr" value="4.5"/> 
        <param name="ktt" value="20.0"/> 
        <param name="ktt2" value="30.0"/>
        <!--  SOME USEFUL RULES: kp = wn*wn  AND kv = 2 xsi wn--> 
        <param name="kv" value="1.0"/> 
        <param name="kp" value="1.4"/> 
        <param name="sigma_p" value="0.5"/> 
        <param name="sigma_v" value="0.5"/> 

        <param name="throttle_neutral" value="1450"/> 
        <param name="quadrotor_mass" value="1.442"/>
        <!-- <param name="quadrotor_mass" value="1.669"/> -->

        <!-- Simulator Node -->
<!--         <node pkg="quad_control" name="Simulator_Iris1" type="quad_simulator.py" output="screen"/>
 -->
        <!-- Simulator parameters -->
        <param name="gravity_sim" value="9.81"/>
        <param name="mass_quad_sim" value="1.442"/>
        <param name="Throttle_neutral_sim" value="1484.0"/>  
        <param name="ACRO_RP_P_sim" value="4.5"/> 
        <param name="ktt_inner_loop" value="10.0"/> 

        <!-- Planner parameters -->
        <param name="D_OPT" value="3" type="double"/> 
        <param name="planner_frequency" value="1e1" type="double"/> 
        <!-- Collision avoidance parameters -->
        <param name="collision_avoidance_active" value="True" type="bool"/>
        <param name="smooth" value="True" type="bool"/>

        <!-- Real arena -->
        <param name="x_min" value="-1.5" type="double"/>
        <param name="y_min" value="-2.2" type="double"/>
        <param name="z_min" value="0.3" type="double"/>
        <param name="x_max" value="2" type="double"/>
        <param name="y_max" value="1.7" type="double"/>
        <param name="z_max" value="3" type="double"/>

        <param name="delta_walls_x" value="0" type="double"/>
        <param name="delta_walls_y" value="0" type="double"/>
        <param name="delta_walls_z" value="0" type="double"/>
        <param name="delta_others" value="2" type="double"/>
        <param name="delta_worry" value="0.5" type="double"/>
        <!-- Magnitude control parameters -->
        <param name="magnitude_control_type" value="BLS" type="String"/>
        <param name="v_lim" value="0.3" type="double"/>
        <param name="v_min" value="1e-2" type="double"/>
        <param name="omega_lim" value="0.79" type="double"/>
        <param name="omega_min" value="0.1" type="double"/>
        <param name="alpha" value="1" type="double"/>
        <param name="beta" value="0.5" type="double"/>
        <param name="tau" value="0.9" type="double"/>

        <param name="buffer_length" value="50" type="int"/>

        <!-- Multiagent mission parameters -->
        <param name="name_others" value="" type="str"/>

        <param name="trade_allowed" value="False" type="bool"/>


        <!-- PARAMETERS THAT ARE COMMOM BETWEEN CONTROLLER AND SIMULATOR -->
        <!-- see WEB: ARDUCOPTER DIFFERENT MODES PARAMETERS -->
        <param name="MAX_ANGLE_DEG" value="45.0"/> 
        <param name="MAX_PSI_SPEED_Deg" value="200.0"/> 

        <!-- MAVROS -->
        <!-- In order to run mavros node, you need to have access to device connected to ttyUSB0, so you may have to change some permissions -->

        
        <arg name="fcu_url" default="/dev/ttyUSB0:57600" />
        <arg name="gcs_url" default="" />
        <arg name="tgt_system" default="1" />
        <arg name="tgt_component" default="1" />
        <arg name="log_output" default="screen" />

        <include file="$(find mavros)/launch/node.launch">
            <arg name="pluginlists_yaml" value="$(find mavros)/launch/apm_pluginlists.yaml" />
            <arg name="config_yaml" value="$(find mavros)/launch/apm_config.yaml" />

            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="$(arg gcs_url)" />
            <arg name="tgt_system" value="$(arg tgt_system)" />
            <arg name="tgt_component" value="$(arg tgt_component)" />
            <arg name="log_output" value="$(arg log_output)" />
        </include>  
        

        <!-- GUI -->
        <!-- <node name="rqt_gui" pkg="gui" type="tabbedGUI"/> -->
        <!-- <node name="rqt_gui" pkg="gui" type="tabPlugin"/> -->

        <!-- This runs rqt and we have to select plugin manually -->
        <!-- similar to rosrun rqt_gui rqt_gui -->
        <!-- <node name="rqt_gui" pkg="rqt_gui" type="rqt_gui"/> -->
        <node name="rqt_gui" pkg="rqt_gui" type="rqt_gui" args="--standalone  tabbedGUI"/>

        <node pkg="quad_control" name="vision_plot_node" type="vision_plot.py"/>

        <node pkg="quad_control" name="projector_node" type="projector_node.py"/>

        <param name="radius_circle_proj" value="0.5" type="double"/>

</launch>